nav2_costmap_2d/src/layered_costmap.cpp

```diff
@@ -145,40 +145,46 @@ void LayeredCostmap::updateMap(double robot_x, double robot_y, double robot_yaw)
   minx_ = miny_ = std::numeric_limits<double>::max();
   maxx_ = maxy_ = std::numeric_limits<double>::lowest();
 
-  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin();
-    plugin != plugins_.end(); ++plugin)
+  #pragma omp parallel reduction(min:minx_) reduction(min: miny_) \
+  reduction(max: maxx_) reduction(max: maxy_) num_threads(2)
   {
-    double prev_minx = minx_;
-    double prev_miny = miny_;
-    double prev_maxx = maxx_;
-    double prev_maxy = maxy_;
-    (*plugin)->updateBounds(robot_x, robot_y, robot_yaw, &minx_, &miny_, &maxx_, &maxy_);
-    if (minx_ > prev_minx || miny_ > prev_miny || maxx_ < prev_maxx || maxy_ < prev_maxy) {
-      RCLCPP_WARN(
-        rclcpp::get_logger(
-          "nav2_costmap_2d"), "Illegal bounds change, was [tl: (%f, %f), br: (%f, %f)], but "
-        "is now [tl: (%f, %f), br: (%f, %f)]. The offending layer is %s",
-        prev_minx, prev_miny, prev_maxx, prev_maxy,
-        minx_, miny_, maxx_, maxy_,
-        (*plugin)->getName().c_str());
+  #pragma omp for schedule(static, 1)
+    for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin();
+      plugin != plugins_.end(); ++plugin)
+    {
+      double prev_minx = minx_;
+      double prev_miny = miny_;
+      double prev_maxx = maxx_;
+      double prev_maxy = maxy_;
+      (*plugin)->updateBounds(robot_x, robot_y, robot_yaw, &minx_, &miny_, &maxx_, &maxy_);
+      if (minx_ > prev_minx || miny_ > prev_miny || maxx_ < prev_maxx || maxy_ < prev_maxy) {
+        RCLCPP_WARN(
+          rclcpp::get_logger(
+            "nav2_costmap_2d"), "Illegal bounds change, was [tl: (%f, %f), br: (%f, %f)], but "
+          "is now [tl: (%f, %f), br: (%f, %f)]. The offending layer is %s",
+          prev_minx, prev_miny, prev_maxx, prev_maxy,
+          minx_, miny_, maxx_, maxy_,
+          (*plugin)->getName().c_str());
+      }
     }
-  }
-  for (vector<std::shared_ptr<Layer>>::iterator filter = filters_.begin();
-    filter != filters_.end(); ++filter)
-  {
-    double prev_minx = minx_;
-    double prev_miny = miny_;
-    double prev_maxx = maxx_;
-    double prev_maxy = maxy_;
-    (*filter)->updateBounds(robot_x, robot_y, robot_yaw, &minx_, &miny_, &maxx_, &maxy_);
-    if (minx_ > prev_minx || miny_ > prev_miny || maxx_ < prev_maxx || maxy_ < prev_maxy) {
-      RCLCPP_WARN(
-        rclcpp::get_logger(
-          "nav2_costmap_2d"), "Illegal bounds change, was [tl: (%f, %f), br: (%f, %f)], but "
-        "is now [tl: (%f, %f), br: (%f, %f)]. The offending filter is %s",
-        prev_minx, prev_miny, prev_maxx, prev_maxy,
-        minx_, miny_, maxx_, maxy_,
-        (*filter)->getName().c_str());
+  #pragma omp for nowait schedule(static, 1)
```

Comment:
Why nowait here? Wouldn't the closed bracket ending the openMP section be right after this to not have any thing to do after the first thread finishes anyway?

Reply:
OpenMP implicitly adds a `barrier` construct after each section in this case `for`. Thus if one thread is finished with a plugin update and another one is still doing its thing the first one will wait for it.

Reply:
Why does the other one not have `nowait`?

Reply:
At the beginning of the parallel block, all threads start their job without any hesitation. Code snippet below might be clearer:
```
    #pragma omp parallel
    {
        // all thread start from here, no stopping

        for (size_t i = 0; i < 100; i++)
        {
            /* code */
        }
        
        #pragma omp barier // implicitly added after each for block
        // all threads wait until the first loop is finished

        for (size_t i = 0; i < 100; i++)
        {
            /* code */
        }

        #pragma omp barier // implicitly added after each for block
    }

Reply:
I don't understand. Wouldn't you want it to potentially continue on in this case? Wouldn't that be the value of `nowait`? The second `updateBounds` loop has `nowait` but the first doesn't, I'd think the first is where it would have more value. I don't think I understand why its not in both.

Reply:
Sorry, my mistake. I thought that `nowait` must be added on the following block in order to let loose the previous block. I'll move `nowait` above.